llvm/utils/TableGen/DynamicTableEmitter.cpp

```diff
@@ -0,0 +1,272 @@
+
+  void run(raw_ostream &OS);
+
+private:
+  std::string primaryRepresentation(SMLoc Loc, const GenericField &Field,
+                                    Init *I) {
+    if (StringInit *SI = dyn_cast<StringInit>(I)) {
+      if (Field.IsCode || SI->hasCodeFormat())
+        return std::string(SI->getValue());
+
+      return SI->getAsString();
+    } else if (BitsInit *BI = dyn_cast<BitsInit>(I))
+      return "0x" + utohexstr(getAsInt(BI));
+    else if (BitInit *BI = dyn_cast<BitInit>(I))
+      return BI->getValue() ? "true" : "false";
+    else if (Field.IsList) {
+      if (auto LI = dyn_cast<ListInit>(I)) {
+        std::stringstream res;
+        auto values = LI->getValues();
+        bool IsAspect = (Field.Name == "aspects");
+        for (const auto &[idx, val] : enumerate(values)) {
+          if (idx > 0)
+            res << ", ";
+
+          if (IsAspect) {
+            auto rec = LI->getElementAsRecord(idx);
+            res << "\"" << rec->getValueAsString("Name").str() << "\"";
+          } else
+            res << val->getAsString();
+          }
+          return res.str();
+      }
+      PrintFatalError(Loc,
+                      Twine("Entry for field '") + Field.Name + "' is null");
+    }
+    PrintFatalError(Loc, Twine("invalid field type for field '") + Field.Name + 
+                             "'; expected: bit, bits, string, or code");
+  }
+
+  void emitDynamicTable(const DynamicTable &Table, raw_ostream &OS);
+  void emitIfdef(StringRef Guard, raw_ostream &OS);
+
+  bool parseFieldType(GenericField &Field, Init *II);
+  void collectTableEntries(DynamicTable &Table,
+                           const std::vector<Record *> &Items);
+};
+
+} // End anonymous namespace.
+
+void DynamicTableEmitter::emitIfdef(StringRef Guard, raw_ostream &OS) {
+  OS << "#ifdef " << Guard << "\n";
+  PreprocessorGuards.insert(std::string(Guard));
+}
+
+void DynamicTableEmitter::emitDynamicTable(const DynamicTable &Table,
+                                           raw_ostream &OS) {
+  emitIfdef((Twine("GET_") + Table.PreprocessorGuard + "_IMPL").str(), OS);
+
+  // The primary data table contains all the fields defined for this map.
+  OS << "std::map<std::string, " << Table.CppTypeName << "> " << Table.Name
+     << " = {\n";
+  for (unsigned i = 0; i < Table.Entries.size(); ++i) {
+    Record *Entry = Table.Entries[i];
+    OS << "  { ";
+
+    ListSeparator LS;
+    bool first = true;
+    bool second = false;
+    for (const auto &Field : Table.Fields) {
+      std::string val = primaryRepresentation(Table.Locs[0], Field,
+                                              Entry->getValueInit(Field.Name));
+      if (first) {
+        first = false;
+        second = true;
+        OS << LS << val << ", { ";
+      } else {
+        if (!second) {
+          OS << LS;
+        }
+        if (Field.IsList) {
+          val.insert(0, 1, '{');
+          val.push_back('}');
+        }
+        OS << val;
+        second = false;
```

Comment:
This logic is a bit hard to understand at first glance. I _think_ that there are some possibilities to do refactoring and simplify it, but it is hard to suggest without understanding the output here.

Which brings me to an important point: there are no tests for this new tablegen backend and we should add some. See [`llvm/test/TableGen`](https://github.com/intel/llvm/tree/sycl/llvm/test/TableGen) for examples of existing ones